onnxruntime/core/providers/rocm/roctracer_manager.cc

```diff
@@ -0,0 +1,465 @@
+  if (p_event_list != nullptr) {
+    p_event_list->emplace_back(std::move(event));
+  }
+}
+
+void RoctracerManager::ProcessActivityBuffers(const std::vector<RoctracerActivityBuffer>& buffers,
+                                              const TimePoint& start_time) {
+  InlinedHashMap<uint64_t, std::vector<EventRecord>> events_pending_client_mapping;
+  auto start_time_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(start_time.time_since_epoch()).count();
+
+  for (auto const& buffer : buffers) {
+    auto current_record = reinterpret_cast<const roctracer_record_t*>(buffer.GetData());
+    auto data_end = reinterpret_cast<const roctracer_record_t*>(buffer.GetData() + buffer.GetSize());
+    for (; current_record < data_end; roctracer_next_record(current_record, &current_record)) {
+      EventRecord event;
+      if (current_record->domain == ACTIVITY_DOMAIN_EXT_API) {
+        roctracer_correlation_to_unique_correlation_[current_record->correlation_id] = current_record->external_id;
+
+        // check for any events pending client mapping on this correlation
+        auto pending_it = events_pending_client_mapping.find(current_record->correlation_id);
+        if (pending_it == events_pending_client_mapping.end()) {
+          continue;
+        }
+
+        // we have one or more pending events, map them to the client
+        MapEventsToClient(current_record->external_id, std::move(pending_it->second));
+        events_pending_client_mapping.erase(current_record->correlation_id);
+        // no additional events to be mapped for this record
+        continue;
+      } else if (current_record->domain == ACTIVITY_DOMAIN_HIP_OPS) {
+        if (current_record->op == 1 && current_record->kind == HipOpMarker) {
+          // this is just a marker, ignore it.
+          continue;
+        }
+
+        auto api_it = api_call_args_.find(current_record->correlation_id);
+        if (api_it == api_call_args_.end()) {
+          // we're not tracking this activity, ignore it
+          continue;
+        }
+
+        auto const& call_record = api_it->second;
+        if (!CreateEventForActivityRecord(current_record, start_time_ns, call_record, event)) {
+          // No event created, skip to the next record to avoid associating an empty
+          // event with a client
+          continue;
+        }
+      } else {
+        // ignore the superfluous event: this is probably a HIP API callback, which
+        // we've had to enable to receive external correlation ids
+        continue;
+      }
+
+      // map the event to the right client
+      auto ext_corr_it = roctracer_correlation_to_unique_correlation_.find(current_record->correlation_id);
+      if (ext_corr_it == roctracer_correlation_to_unique_correlation_.end()) {
+        // defer the processing of this event
+        events_pending_client_mapping[current_record->correlation_id].emplace_back(std::move(event));
```

Comment:
Out of curiosity, will it be the case that there will be unmatched events left as residuals after processing the whole passed in `buffers`. But they will have a match if you call this function with (hypothetically) subsequent `buffers`. But since `events_pending_client_mapping` is a local variable, those residuals will be thrown away.

That is, is it guaranteed that the `buffers` tracks all the events in a single profiling session? If it is, will the buffer larger than (some hypothetically) limitation somehow?

Reply:
Great catch, thank you! The right approach is for `events_pending_client_mapping` to be an instance variable, so that we retain any events for which we haven't yet received an external_correlation activity record.